hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py

```diff
@@ -0,0 +1,233 @@
+
+    _assert_loss_cache_contains_n_elements(loss_callback.loss_cache, n_slides_per_process)
+    loss_callback.on_train_epoch_end(trainer, pl_module)
+    # Loss cache is flushed after each epoch
+    _assert_loss_cache_contains_n_elements(loss_callback.loss_cache, 0)
+
+    if rank > 0:
+        time.sleep(10)  # Wait for the rank 0 to save the loss cache in a csv file
+
+    loss_cache_path = loss_callback.cache_folder / LOSS_VALUES_FILENAME.format(loss_callback.zfill_epoch(current_epoch))
+    assert loss_callback.cache_folder.exists()
+    assert loss_cache_path.exists()
+    assert loss_cache_path in loss_callback.cache_folder.iterdir()
+
+    loss_cache = pd.read_csv(loss_cache_path)
+    total_slides = n_slides_per_process * world_size if not duplicate else n_slides_per_process * world_size - 1
+    _assert_loss_cache_contains_n_elements(loss_cache, total_slides)
+    _assert_list_is_sorted(loss_cache[ResultsKey.LOSS].values)
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="Not enough GPUs available")
+@pytest.mark.gpu
+def test_on_train_epoch_end_distributed(tmp_path: Path) -> None:
+    # Test that the loss cache is saved correctly when using multiple GPUs
+    # First scenario: no duplicates
+    run_distributed(test_on_train_epoch_end, [tmp_path, False], world_size=2)
+    # Second scenario: introduce duplicates
+    run_distributed(test_on_train_epoch_end, [tmp_path, True], world_size=2)
+
+
+def test_on_train_end(tmp_path: Path) -> None:
+    trainer = MagicMock()
+    pl_module = MagicMock(global_rank=0)
+    max_epochs = 4
+
+    loss_callback = LossAnalysisCallback(
+        outputs_folder=tmp_path, max_epochs=max_epochs, num_slides_heatmap=2, num_slides_scatter=2
+    )
+    dump_loss_cache_for_epochs(loss_callback, max_epochs)
+    loss_callback.on_train_end(trainer, pl_module)
+
+    for epoch in range(max_epochs):
+        assert (loss_callback.cache_folder / LOSS_VALUES_FILENAME.format(loss_callback.zfill_epoch(epoch))).exists()
+
+    # check save_loss_ranks outputs
+    assert (loss_callback.cache_folder / ALL_EPOCHS_FILENAME).exists()
+    assert (loss_callback.rank_folder / LOSS_STATS_FILENAME).is_file()
+    assert (loss_callback.rank_folder / LOSS_RANKS_FILENAME).exists()
+    assert (loss_callback.rank_folder / LOSS_RANKS_STATS_FILENAME).exists()
+
+    # check plot_slides_loss_scatter outputs
+    assert (loss_callback.scatter_folder / SCATTER_PLOT_FILENAME.format(HIGHEST)).exists()
+    assert (loss_callback.scatter_folder / SCATTER_PLOT_FILENAME.format(LOWEST)).exists()
+
+    # check plot_loss_heatmap_for_slides_of_epoch outputs
+    for epoch in range(max_epochs):
+        filename = HEATMAP_PLOT_FILENAME.format(loss_callback.zfill_epoch(epoch), HIGHEST)
+        assert (loss_callback.heatmap_folder / filename).exists()
+        filename = HEATMAP_PLOT_FILENAME.format(loss_callback.zfill_epoch(epoch), LOWEST)
+        assert (loss_callback.heatmap_folder / filename).exists()
```

Comment:
Consider creating convenience methods to generate all these file paths (especially the ones with variables), for readability and to ensure they're always formatted the same way.

Reply:
In my opinion, the API would be cleaner with a separate method to get the path for each of these files, so the user doesn't have to remember which folder and parameters need to be provided for each of them. E.g.
- `get_all_epochs_file()`
- `get_loss_values_file(epoch)`
- `get_heatmap_plot_file(epoch, order)`
- etc.

What do you think?